sdk/src/settings.rs

```diff
@@ -0,0 +1,581 @@
+
+            if let Ok(cloned_config) = cloned_config {
+                match cloned_config.try_deserialize::<Settings>() {
+                    Ok(s) => Some(s),
+                    Err(_) => None,
+                }
+            } else {
+                None
+            }
+        }
+        Err(_) => None,
+    }
+}
+
+// Load settings from configuration file
+#[allow(unused)]
+pub fn load_settings<P: AsRef<Path>>(settings_path: P) -> Result<()> {
+    let ext = settings_path
+        .as_ref()
+        .extension()
+        .ok_or(Error::UnsupportedType)?
+        .to_string_lossy();
+
+    let setting_buf = std::fs::read(&settings_path).map_err(Error::IoError)?;
+
+    load_settings_from_str(&String::from_utf8_lossy(&setting_buf), &ext)
+}
+
+// Load settings form string representation of the configuration.  Format of configuration must be supplied.
+#[allow(unused)]
+pub fn load_settings_from_str(settings_str: &str, format: &str) -> Result<()> {
+    Settings::from_string(settings_str, format).map(|_| ())
+}
+
+// Save the current configuration to a json file.
+#[allow(unused)]
+pub fn save_settings_as_json<P: AsRef<Path>>(settings_path: P) -> Result<()> {
+    let settings =
+        get_settings().ok_or(Error::OtherError("could not get current settings".into()))?;
+
+    let settings_json = serde_json::to_string_pretty(&settings).map_err(Error::JsonError)?;
+
+    std::fs::write(settings_path, settings_json.as_bytes()).map_err(Error::IoError)
+}
+
+// Set a Settings value by path reference.  The path is nested names of of the Settings objects
+// separated by "." notation.  For example "core.hash_alg" would set settings.core.hash_alg value.
+// The nesting can be arbitrarily deep based on the Settings definition.
+#[allow(unused)]
+pub(crate) fn set_settings_value<T: Into<config::Value>>(value_path: &str, value: T) -> Result<()> {
+    match SETTINGS.write() {
+        Ok(mut c) => {
+            let source = c.clone();
+            let update_config = Config::builder()
+                .add_source(source)
+                .set_override(value_path, value);
+
+            if let Ok(updated) = update_config {
+                *c = updated
+                    .build()
+                    .map_err(|_e| Error::OtherError("could not update configuration".into()))?;
+                Ok(())
+            } else {
+                Err(Error::OtherError("could not save settings".into()))
+            }
+        }
+        Err(_) => Err(Error::OtherError("could not save settings".into())),
+    }
+}
+
+// Get a Settings value by path reference.  The path is nested names of of the Settings objects
+// separated by "." notation.  For example "core.hash_alg" would get the settings.core.hash_alg value.
+// The nesting can be arbitrarily deep based on the Settings definition.
+#[allow(unused)]
+pub(crate) fn get_settings_value<'de, T: serde::de::Deserialize<'de>>(
+    value_path: &str,
+) -> Option<T> {
```

Comment:
This one kind of worries me because an error in the deserialization will only be seen as None. This may be a good case for a Return<Option<T>> But I don't know. We could also add some logging to at least detect somehow that something went wrong here.

Reply:
Deserialization error should never happen because the values are checked before saving the configuration.